lib/spack/docs/packaging_guide.rst

```diff
@@ -2016,45 +2016,124 @@ the Python extensions provided by them: once for ``+python`` and once
+
+When a user runs ``spack install``, Spack:
+
+* fetches an archive for the correct version of the software
+* expands the archive
+* sets the current working directory to the root directory of the expanded archive
+
+Then, depending on the base class of the package under consideration, it will execute
+a certain number of **phases** that reflect the way a package of that type is usually built.
+The name and order in which the phases will be executed can be obtained either reading the API docs in this manual, or
+using the ``spack info`` command:
+
+.. code-block:: console
+    :emphasize-lines: 13,14
+
+    $ spack info m4
+    AutotoolsPackage:    m4
+    Homepage:            https://www.gnu.org/software/m4/m4.html
+
+    Safe versions:
+        1.4.17    ftp://ftp.gnu.org/gnu/m4/m4-1.4.17.tar.gz
+
+    Variants:
+        Name       Default   Description
+
+        sigsegv    on        Build the libsigsegv dependency
+
+    Installation Phases:
+        autoreconf    configure    build    install
+
+    Build Dependencies:
+        libsigsegv
 
-``install`` takes a ``spec``: a description of how the package should
-be built, and a ``prefix``: the path to the directory where the
-software should be installed.
+    Link Dependencies:
+        libsigsegv
 
-Spack provides wrapper functions for ``configure`` and ``make`` so
-that you can call them in a similar way to how you'd call a shell
-command.  In reality, these are Python functions.  Spack provides
-these functions to make writing packages more natural. See the section
-on :ref:`shell wrappers <shell-wrappers>`.
+    Run Dependencies:
+        None
 
-Now that the metadata is out of the way, we can move on to the
-``install()`` method.  When a user runs ``spack install``, Spack
-fetches an archive for the correct version of the software, expands
-the archive, and sets the current working directory to the root
-directory of the expanded archive.  It then instantiates a package
-object and calls the ``install()`` method.
+    Virtual Packages:
+        None
 
-The ``install()`` signature looks like this:
+    Description:
+        GNU M4 is an implementation of the traditional Unix macro processor.
+
+Typically, phases have  default implementations that fit most of the common cases:
+
+.. literalinclude:: ../../../lib/spack/spack/build_systems/autotools.py
+    :pyobject: AutotoolsPackage.configure
+    :linenos:
+
+making it just sufficient for a packager to override a few
+build-system specific helper methods or attributes to provide, for instance,
+configure arguments:
+
+.. literalinclude::  ../../../var/spack/repos/builtin/packages/m4/package.py
+    :pyobject: M4.configure_args
+    :linenos:
+
+.. note::
+    Each specific build-system has a list of attributes that can be overridden to
+    fine-tune the installation of a package without overriding an entire phase. To
+    have more information on them the place to go is the API docs at the end of this manual.
```

Comment:
Should include a link here to build system-specific sections of the manual: CMakePackage, PythonPackage, etc.  (I see that such sections are below in this PR, in the API docs).

I think we also need at least 1-2 examples that "put it all together".  We should see cases where `configure_args()` or `cmake_args()` is used, and where entire phases are overridden.  That would go in a tutorial section, this looks more like a reference manual section.